### PR TITLE
support for fillOpacity in swatch and ramp color legends

### LINKED ISSUE #1356: color legend doesn't account for fillOpacity
Color legends show the color with opacity: 1. Thus the legend might look quite different than the plot, in case you set the fillOpacity < 1

### PR DESCRIPTION
**Pull request by Fil**
Note: I used **fillOpacity** and not **opacity**, for consistency with the *symbol* legends.

closes #1356

### CODE REVIEW

**Review by mbostock — CHANGES_REQUESTED** [review]
The reason I suggested calling it **opacity** (and not **fillOpacity**) is that we use the **color** option for the _opacity_ _ramp_ legend. Also more practically, there is no possible stroke here, so we do not need to distinguish between fill and stroke, and opacity is shorter.

**Comment by Fil** [comment]
My code was really missing a few pieces :-(

Now, it's a bummer that we have to select **opacity** if there is no symbol, **fillOpacity** if the symbols are filled, and **strokeOpacity** if they are stroked. Maybe we could support opacity across the board (as well as strokeOpacity and fillOpacity)?

File: src/legends/swatches.js [hunk 1]
```diff
@@ -12,14 +12,20 @@ function maybeScale(scale, key) {
...
     options,
     (selection, scale) =>
-      selection.append("svg").attr("fill", scale.scale).append("rect").attr("width", "100%").attr("height", "100%"),
+      selection
+        .append("svg")
+        .attr("fill", scale.scale)
+        .append("rect")
+        .attr("width", "100%")
+        .attr("height", "100%")
+        .attr("opacity", maybeNumberChannel(opacity)[1]),
```
Comment (mbostock) [inline src/legends/swatches.js:28]: We can implement the option as fill-opacity here (for faster rendering, not that it is likely to matter in this case).

```suggestion
        .attr("fill-opacity", maybeNumberChannel(opacity)[1]),
```

File: README.md [hunk 2]
```diff
@@ -679,6 +680,7 @@ Continuous color legends are rendered as a ramp, and can be configured with the
 * *options*.**marginRight** - the legend’s right margin
 * *options*.**marginBottom** - the legend’s bottom margin
 * *options*.**marginLeft** - the legend’s left margin
+* *options*.**opacity** - the ramp’s opacity
```
Comment (mbostock) [inline README.md:683]: Did we forget to document the color option for opacity ramp legends?

File: src/legends/ramp.js [hunk 3]
```diff
@@ -117,6 +121,7 @@ export function legendRamp(color, options) {
 
     svg
       .append("g")
+      .attr("opacity", opacity)
```
Comment (mbostock) [inline src/legends/ramp.js:124]: Implement as fill-opacity for performance?

```suggestion
      .attr("fill-opacity", opacity)
```

File: src/legends/ramp.js [hunk 4]
```diff
@@ -137,6 +142,7 @@ export function legendRamp(color, options) {
...
```
Comment (mbostock) [inline src/legends/ramp.js:145]: Likewise…

```suggestion
      .attr("fill-opacity", opacity)
```

File: src/legends/swatches.js (hunk 1 above)
Comment (Fil) [reply src/legends/swatches.js:28]: yup, why not; but then maybe also document that (by adding "fill" in README and in the jsdocs?)

Comment (mbostock) [reply src/legends/swatches.js:28]: I’m not sure we need to document it because it’s an implementation detail how the swatches are rendered… but sure.

**Review by mbostock — APPROVED** [review]

**Pull request merged**